src/utils/pgwire/src/pg_protocol.rs

```diff
@@ -127,30 +150,101 @@ where
+                    unnamed_statement
+                } else {
+                    // NOTE Error handle method may need to modified
+                    named_statements.get(&statement_name).unwrap()
+                };
+                // Step 2 instance
+                let portal_name = cstr_to_str(&m.portal_name).unwrap().to_string();
+                let portal = statement.instance(portal_name.clone(), &m.params);
+                // Step 3 Store Portal
+                if portal_name.is_empty() {
+                    *unnamed_portal = portal;
+                } else {
+                    named_portals.insert(portal_name, portal);
+                }
                 self.write_message(&BeMessage::BindComplete).await?;
             }
-            FeMessage::Execute(_) => {
-                self.process_query_msg(cstr_to_str(unnamed_query_string), true)
+            FeMessage::Execute(m) => {
+                // Step 1 Get portal
+                let portal_name = cstr_to_str(&m.portal_name).unwrap().to_string();
+                let portal = if m.portal_name.is_empty() {
+                    unnamed_portal
+                } else {
+                    // NOTE: error handle need modify later;
+                    named_portals.get(&portal_name).unwrap()
+                };
+                // Step 2 Execute instance statement using portal
+                self.process_query_msg(cstr_to_str(&portal.get_query_string()), true)
                     .await?;
                 // NOTE there is no ReadyForQuery message.
             }
-            FeMessage::Describe(_) => {
-                self.write_message_no_flush(&BeMessage::ParameterDescription)?;
+            FeMessage::Describe(m) => {
                 // FIXME: Introduce parser to analyze statements and bind data type. Here just
                 // hard-code a VARCHAR.
-                self.write_message(&BeMessage::RowDescription(&[PgFieldDescriptor::new(
-                    "VARCHAR".to_string(),
-                    TypeOid::Varchar,
-                )]))
-                .await?;
+                // Step 1 Get statement
+                let name = cstr_to_str(&m.query_name).unwrap().to_string();
+                let statement = if name.is_empty() {
+                    unnamed_statement
+                } else {
+                    // NOTE: error handle need modify later;
+                    named_statements.get(&name).unwrap()
+                };
+                // Step 2 Send parameter description
+                self.write_message(&BeMessage::ParameterDescription(&statement.get_type_desc()))
+                    .await?;
+                // Step 3 Send row description
+                self.write_message(&BeMessage::RowDescription(&statement.get_row_desc()))
+                    .await?;
             }
             FeMessage::Sync => {
                 self.write_message(&BeMessage::ReadyForQuery).await?;
             }
+            FeMessage::Close(m) => {
```

Comment:
So close will be called when `transaction.rollback()`?

Reply:
When the 'statement' or 'portal' object in tokio_postgres drops, their drop function will send the 'close' msg to the server to ask to delete the corresponding prepare_statement or portal.